File: solutions/314. Binary Tree Vertical Order Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    // soln by ldytmac
private: 
    // first int : col of this node, second int : row of this node
    // vector<int> : stores vals of nodes at this (row, col), use vector because may have 2 nodes of the same (row, col)
    map<int, map<int, vector<int>>> m;
    void helper(TreeNode* node, int row, int col) {
        if(!node) return;
        m[col][row].push_back(node->val);
        // this order make sure if two nodes are in the same row and column, the order should be from left to right.
        if(node->left) helper(node->left, row+1, col-1);
        if(node->right) helper(node->right, row+1, col+1);        
    }
public:
    vector<vector<int>> verticalOrder(TreeNode* root) {
        helper(root, 0, 0);
        vector<vector<int>> res;
        // add vals of each col
        for (auto colItr = m.begin(); colItr != m.end(); ++colItr) {
            vector<int> tmp;
            // add vals of each (row, col)
            for (auto rowItr = colItr->second.begin(); rowItr != colItr->second.end(); rowItr++) {
                for(int& val : rowItr->second) tmp.push_back(val);
            }
            res.push_back(tmp);
        }
        return res;
    }
};

File: solutions/314. Binary Tree Vertical Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def verticalOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
#         # Leetcode soln 2, time O(N), space O(N)
#         if root is None:
#             return []
        
#         max_col = min_col = 0
#         column_lookup = defaultdict(list)
#         queue = deque([(root,0)])
        
#         while queue:
#             node, col = queue.popleft()
#             if node is not None:
#                 max_col = max(max_col, col)
#                 min_col = min(min_col, col)
#                 column_lookup[col].append(node.val)
                
#                 queue.append((node.left, col-1))
#                 queue.append((node.right, col+1))
                
#         return [column_lookup[k] for k in range(min_col, max_col+1)]      
        
#         # Leetcode soln 1, time O(NlogN), space O(N)
        column_lookup = defaultdict(list)
        queue = deque([(root,0)])
        
        while queue:
            node, col = queue.popleft()
            if node is not None:
                column_lookup[col].append(node.val)
                
                queue.append((node.left, col-1))
                queue.append((node.right, col+1))
                
        return [column_lookup[k] for k in sorted(column_lookup.keys())]
    
    
